File: ImperatorToCK3/Source/Mappers/RegionMapper/ImperatorArea.cpp
#include "ImperatorArea.h"
#include "ParserHelpers.h"
#include "CommonRegexes.h"
#include "Log.h"

mappers::ImperatorArea::ImperatorArea(std::istream& theStream)
{
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}

void mappers::ImperatorArea::registerKeys()
{
	registerKeyword("provinces", [this](std::istream& theStream) {
		for (const auto& id : commonItems::getULlongs(theStream))
			provinces.insert(id);
	});
	registerMatcher(commonItems::catchallRegexMatch, commonItems::ignoreItem);
}

bool mappers::ImperatorArea::areaContainsProvince(const unsigned long long province) const
{
	return provinces.contains(province);
}

File: ImperatorToCK3/Mappers/Region/ImperatorArea.cs
using commonItems;
using commonItems.Collections;
using System.Collections.Generic;

namespace ImperatorToCK3.Mappers.Region;

public class ImperatorArea : IIdentifiable<string> {
	public SortedSet<ulong> ProvinceIds { get; } = new();
	public string Id { get; }

	public ImperatorArea(string id, BufferedReader reader) {
		Id = id;
		var parser = new Parser();
		RegisterKeys(parser);
		parser.ParseStream(reader);
	}
	private void RegisterKeys(Parser parser) {
		parser.RegisterKeyword("provinces", reader => ProvinceIds.UnionWith(reader.GetULongs()));
		parser.IgnoreAndLogUnregisteredItems();
	}
	public bool ContainsProvince(ulong province) {
		return ProvinceIds.Contains(province);
	}
}